File: MetaheuristicOptimization/AllMetaheuristic/Nature-inspired/swarm-based/Buffalo Swarm Optimization_re/buffelows/buffalo_swarm_optimization.m
% Buffalo Swarm Optimization (BSO) for Sphere Function

function [best_solution, best_fitness] = buffalo_swarm_optimization(num_buffaloes, num_iterations, num_variables, lower_bound, upper_bound)
    % Initialize buffalo population
    buffaloes = rand(num_buffaloes, num_variables) .* (upper_bound - lower_bound) + lower_bound;
    
    % Define objective function (sphere function)
    objective_function = @(x) sum((max(min(x, upper_bound), lower_bound)).^2); % Sphere function with bounds
    
    % Initialize best solution and best fitness
    best_solution = buffaloes(1,:);
    best_fitness = objective_function(best_solution);
    
    % Main loop
    for iter = 1:num_iterations
        % Update each buffalo's position
        for i = 1:num_buffaloes
            % Calculate new position based on local search
            new_position = local_search_with_bounds(buffaloes(i,:), objective_function, lower_bound, upper_bound);
            
            % Update buffalo position
            buffaloes(i,:) = new_position;
            
            % Update best solution and best fitness if needed
            fitness = objective_function(new_position);
            if fitness < best_fitness
                best_fitness = fitness;
                best_solution = new_position;
            end
        end
    end
end

% Local search function with bounds
function new_position = local_search_with_bounds(current_position, objective_function, lower_bound, upper_bound)
    % Perturb current position
    perturbation = randn(size(current_position)) * 0.1; % Small random perturbation
    new_position = current_position + perturbation;
    
    % Ensure new position is within bounds
    new_position = max(min(new_position, upper_bound), lower_bound);
end

File: documentation/AllMetaheuristic/Nature-inspired/swarm-based/buffalo_swarm_optimization.py
import numpy as np

def buffalo_swarm_optimization(objective_function, num_buffaloes, num_iterations, num_variables, lower_bound, upper_bound):
    """
    Buffalo Swarm Optimization (BSO) for a given objective function.

    Parameters:
    - objective_function: Function to optimize (takes a numpy array and returns a scalar).
    - num_buffaloes: Number of buffaloes in the population.
    - num_iterations: Maximum number of iterations.
    - num_variables: Number of dimensions (variables).
    - lower_bound: Lower bound for each dimension.
    - upper_bound: Upper bound for each dimension.

    Returns:
    - best_solution: Best solution found.
    - best_fitness: Best fitness value.
    - history: List of tuples (iteration, best_solution, best_fitness) for each iteration.
    """
    # Initialize buffalo population
    buffaloes = np.random.rand(num_buffaloes, num_variables) * (upper_bound - lower_bound) + lower_bound
    
    # Initialize best solution and best fitness
    best_solution = buffaloes[0, :].copy()
    best_fitness = objective_function(best_solution)
    
    # Initialize history
    history = [(0, best_solution.copy(), best_fitness)]
    
    # Main loop
    for iter in range(num_iterations):
        # Update each buffalo's position
        for i in range(num_buffaloes):
            # Calculate new position based on local search
            new_position = local_search_with_bounds(buffaloes[i, :], objective_function, lower_bound, upper_bound)
            
            # Update buffalo position
            buffaloes[i, :] = new_position
            
            # Update best solution and best fitness if needed
            fitness = objective_function(new_position)
            if fitness < best_fitness:
                best_fitness = fitness
                best_solution = new_position.copy()
        
        # Store best solution for this iteration
        history.append((iter + 1, best_solution.copy(), best_fitness))
    
    return best_solution, best_fitness, history

def local_search_with_bounds(current_position, objective_function, lower_bound, upper_bound):
    """
    Local search function with bounds.

    Parameters:
    - current_position: Current position of the buffalo.
    - objective_function: Function to evaluate fitness.
    - lower_bound: Lower bound for each dimension.
    - upper_bound: Upper bound for each dimension.

    Returns:
    - new_position: New position after local search.
    """
    # Perturb current position
    perturbation = np.random.randn(len(current_position)) * 0.1  # Small random perturbation
    new_position = current_position + perturbation
    
    # Ensure new position is within bounds
    new_position = np.clip(new_position, lower_bound, upper_bound)
    
    return new_position
